src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/MongoStockRecordRepository.kt

```diff
@@ -131,25 +140,31 @@ class MongoStockRecordRepository(
         field: String,
         from: Date,
         to: Date,
-    ): BigDecimal? {
-        val collection: MongoCollection<Document> = mongoTemplate
-            .getCollection(MongoStockRecord.COLLECTION_NAME)
-
-        return collection
-            .find()
-            .filter(gte(MongoStockRecord::dateOfRetrieval.name, from))
-            .filter(lt(MongoStockRecord::dateOfRetrieval.name, to))
-            .projection(
-                fields(
-                    include(field),
-                    excludeId()
-                )
-            )
-            .sort((Sorts.descending(field)))
-            .limit(1)
-            .firstOrNull()?.let {
-                (it[field] as? Decimal128)?.bigDecimalValue()
-            }
+    ): Mono<BigDecimal> {
+        val matchDateRange = Aggregation.match(
+            Criteria.where(MongoStockRecord::dateOfRetrieval.name)
+                .gte(from)
+                .lt(to)
+        )
+
+        val matchNonNullField = Aggregation.match(
+            Criteria.where(field).ne(null)
+        )
+
+        val project = Aggregation.project(field)
+            .andExclude("_id")
+            .and(field).`as`("max")
+
+        val sort = Aggregation.sort(Sort.Direction.DESC, field)
```

Comment:
sort could be expensive. Can we use something like:
```
db.products.aggregate([
  {
    $group: {
      _id: null, // Grouping by null gives a single result for the entire collection
      maxPrice: { $max: "$price" } // Calculate max of the "price" field
    }
  }
])
```

(The code was generated by AI)